sql/core/src/main/scala/org/apache/spark/sql/sources/interfaces.scala

```diff
@@ -604,10 +609,33 @@ abstract class HadoopFsRelation private[sql](maybePartitionSpec: Option[Partitio
       }
     }
 
-    buildInternalScan(requiredColumns, filters, inputStatuses, broadcastedConf)
+    if (!inputExists) {
+      throw new IOException("Input paths do not exist, input paths="
+        + inputPaths.mkString("[", ",", "]"))
+    } else {
+      if (inputStatuses.isEmpty && readFromHDFS) {
+        logWarning("Input paths are empty, input paths=" + inputPaths.mkString("[", ",", "]"))
+        sqlContext.sparkContext.emptyRDD[InternalRow]
+      } else {
+        buildInternalScan(requiredColumns, filters, inputStatuses, broadcastedConf)
+      }
+    }
   }
 
   /**
+   * Most of time, HadoopFsRelation should check the inputPaths, but for some cases it is not,
+   * e.g. JsonRelation may read from RDD[String]
+   */
+  def inputExists: Boolean = fileStatusCache.inputExists
+
+  /**
+   * Most of time, HadoopFsRelation should read from hdfs, but some cases it is not,
+   * e.g. JsonRelation may read from RDD[String]
+   * @return
+   */
+  def readFromHDFS: Boolean = true
```

Comment:
@yhuai  @liancheng This issue exist also in TextRelation, not only JsonRelation. So it would be better to check the input paths in HadoopFsRelation. But JsonRelation also can accept RDD[String] as input, so I think it would be better to separate JsonRelation that accept RDD[String] to a new XXXRelation. So here's what I proposal:
* Checking input paths in HadoopFsRelation, but don't change its interface
* Separate JsonRelation that accept RDD[String] into a new JsonRDDRelation

What do you think ?